worker/peergrouper/desired.go

```diff
@@ -397,88 +399,128 @@ func (p *peerGroupChanges) getMachinesVoting() {
+			continue
+		}
 
-		if member.Address != addr {
-			member.Address = addr
-			p.isChanged = true
+		// Unique address; we can use this for Mongo peer communication.
+		member := p.members[id]
+		if len(addrs) == 1 {
+			addr := addrs[0]
+			logger.Debugf("machine %q selected address %q by scope from %v", id, addr, hostPorts)
+
+			if member.Address != addr {
+				member.Address = addr
+				p.isChanged = true
+			}
+			continue
+		}
+
+		// Multiple potential Mongo addresses.
+		// Checks are required in order to use it as a peer.
+		unchanged := false
+		if _, ok := info.recognised[id]; ok {
+			for _, addr := range addrs {
+				if member.Address == addr {
+					logger.Warningf("%s\npreserving member with unchanged address %q", multiAddressMessage, addr)
+					unchanged = true
+					break
+				}
+			}
 		}
-		return nil
-	}
 
-	msg := fmt.Sprintf(
-		"machine %q has more than one usable address and juju-ha-space is not set"+
-			"\nrun \"juju config juju-ha-space=<name>\" to set a space for Mongo peer communication", id)
+		// If this member was not previously in the replica-set, or if its
+		// address has changed, we enforce the policy of requiring a
+		// configured HA space when there are multiple cloud-local addresses.
+		if !unchanged {
+			multipleAddresses = append(multipleAddresses, id)
+			if err := m.stm.SetStatus(getStatusInfo(multiAddressMessage)); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
 
-	// If this would-be member is new, we enforce the policy of having a
-	// configured HA space when there are multiple cloud-local addresses.
-	if _, ok := info.recognised[id]; !ok {
-		return errors.New(msg)
+	if len(multipleAddresses) > 0 {
+		ids := strings.Join(multipleAddresses, ", ")
+		return fmt.Errorf("juju-ha-space is not set and these machines have more than one usable address: %s"+
+			"\nrun \"juju config juju-ha-space=<name>\" to set a space for Mongo peer communication", ids)
 	}
+	return nil
+}
+
+// updateAddressesFromSpace updates the member addresses based on the
+// configured HA space.
+// If no addresses are available for any of the machines, then such machines
+// have their status set and are included in the detail of the returned error.
+func (p *peerGroupChanges) updateAddressesFromSpace(info *peerGroupInfo) error {
+	space := info.haSpace
+	var noAddresses []string
 
-	// If the prior address is still available, we allow preservation of the
-	// status quo rather than throwing out with an error.
-	for _, addr := range addrs {
-		if member.Address == addr {
-			logger.Warningf("%s\npreserving member with unchanged address %q", msg, addr)
-			return nil
+	for id := range p.members {
+		m := info.machines[id]
+		addr, err := m.SelectMongoAddressFromSpace(info.mongoPort, space)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if addr == "" {
+			noAddresses = append(noAddresses, id)
+			if err := m.stm.SetStatus(getStatusInfo(fmt.Sprintf("no addresses in space %q", space))); err != nil {
```

Comment:
does "no address found for ha-space=%q" lead people to figuring out the configuration is wrong?
I'm ok with the message, just wondering what breadcrumbs we can leave around.